drivers/sensor/qdec_stm32/qdec_stm32.c

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright (c) 2022, Valerio Setti <vsetti@baylibre.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT st_stm32_qdec
+
+/** @file
+ * @brief STM32 family Quadrature Decoder (QDEC) driver.
+ */
+
+#include <errno.h>
+#include <zephyr/sys/__assert.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/device.h>
+#include <zephyr/init.h>
+#include <soc.h>
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/drivers/clock_control/stm32_clock_control.h>
+
+#include <stm32_ll_tim.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(qdec_stm32, CONFIG_SENSOR_LOG_LEVEL);
+
+/* Device constant configuration parameters */
+struct qdec_stm32_dev_cfg {
+	const struct pinctrl_dev_config *pin_config;
+	struct stm32_pclken pclken;
+	TIM_TypeDef *timer_inst;
+	uint16_t prescaler;
+	uint8_t is_input_polarity_inverted;
+	uint8_t input_filtering_level;
+};
+
+/* Device run time data */
+struct qdec_stm32_dev_data {
+	int32_t position;
+};
+
+static int qdec_stm32_fetch(const struct device *dev, enum sensor_channel chan)
+{
+	struct qdec_stm32_dev_data *dev_data = dev->data;
+	const struct qdec_stm32_dev_cfg *dev_cfg = dev->config;
+
+	dev_data->position = (int32_t) LL_TIM_GetCounter(dev_cfg->timer_inst);
+
+	return 0;
+}
+
+static int qdec_stm32_get(const struct device *dev, enum sensor_channel chan,
+			struct sensor_value *val)
+{
+	struct qdec_stm32_dev_data *const dev_data = dev->data;
+
+	if (chan == SENSOR_CHAN_ROTATION) {
+		val->val1 = dev_data->position;
+		val->val2 = 0;
+	} else {
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static int qdec_stm32_initialize(const struct device *dev)
+{
+	__ASSERT_NO_MSG(dev != NULL);
+	const struct qdec_stm32_dev_cfg *const dev_cfg = dev->config;
+	int retval;
+	LL_TIM_ENCODER_InitTypeDef init_props;
+
+	retval = pinctrl_apply_state(dev_cfg->pin_config, PINCTRL_STATE_DEFAULT);
+	if (retval < 0) {
+		return retval;
+	}
+
+	retval = clock_control_on(DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE),
```

Comment:
Check device is ready before using. See https://github.com/zephyrproject-rtos/zephyr/pull/48787